### PR TITLE
fix: only train for one batch in PyTorch Trainer test mode

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

previous change to enable training lengths longer than searcher length introduced bug where test_mode flag will train for max_length instead of just one batch.

## Test Plan

run a script using pytorch trainer API (example for mnist):
```
import determined as det
import yaml
import logging
from model_def import MNistTrial

def main():

    with open("const.yaml", "r") as file:
        exp_conf = yaml.safe_load(file)

    hparams = exp_conf["hyperparameters"]

    with det.pytorch.init(hparams=hparams, exp_conf=exp_conf) as train_context:
        trial = MNistTrial(train_context)
        trainer = det.pytorch.Trainer(trial, train_context)

        trainer.fit(
            max_length=det.pytorch.Epoch(1),
            test_mode=True
        )

if __name__ == "__main__":
    # Configure logging
    logging.basicConfig(level=logging.INFO, format=det.LOG_FORMAT)
    main()
```

verify in stdout that the code trains for 1 batch only.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py [hunk 1]
```diff
@@ -751,14 +751,14 @@ def _train_for_op(
...
         if self.local_training:
-            searcher_length = self.max_length
+            train_length = Batch(1) if self.test_mode else self.max_length
         else:
-            searcher_length = TrainUnit._from_searcher_unit(
+            train_length = TrainUnit._from_searcher_unit(
                 op.length, self.searcher_unit, self.global_batch_size
             )
-        assert searcher_length
+        assert train_length
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:759]: just confirming, we don't support `--test` without `--local` through any codepath except the legacy codepath, right?

This seems correct under that condition, but might make more sense to check for test mode first, then local, since if we did support nonlocal test mode we would still only train one batch.

Comment (azhou-determined) [reply harness/determined/pytorch/_pytorch_trial.py]: good point. updated.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -751,14 +751,16 @@ def _train_for_op(
...
-            searcher_length = self.max_length
+            train_length = self.max_length
         else:
-            searcher_length = TrainUnit._from_searcher_unit(
+            train_length = TrainUnit._from_searcher_unit(
                 op.length, self.searcher_unit, self.global_batch_size
             )
-        assert searcher_length
+        if self.test_mode:
+            train_length = Batch(1)
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:760]: nit: reorder your if statements so you aren't mutating variables; that would feel a lot more natural:
```python
if self.test_mode:
  train_length = Batch(1)
elif self.local_training:
  train_length = self.max_length
else:
  train_length = TrainUnit._from_searcher_unit(...)
```

Comment (azhou-determined) [reply harness/determined/pytorch/_pytorch_trial.py]: duh. thanks.

**Review by rb-determined-ai — APPROVED** [review]
